### PR TITLE
Specify sequence number endianness

### LINKED ISSUE #2166: TLS1.3 compliance issue w.r.t sequence number handling (AEAD encryption and decryption)
Hi,

Libspdm does not handle sequence number in correct byte order, as per the spec it's encoded in network byte order so it should be handled in the network byte order.

As per TLS1.3 spec section 5.3 (below)

https://datatracker.ietf.org/doc/html/rfc8446#section-5.3

"
The per-record nonce for the AEAD
   construction is formed as follows:

   1.  The 64-bit record sequence number is encoded in network byte
       order and padded to the left with zeros to iv_length.

   2.  The padded sequence number is XORed with either the static
       client_write_iv or server_write_iv (depending on the role).

   The resulting quantity (of length iv_length) is used as the
   per-record nonce.
"

Libspdm reference code -  (file: libspdm_secmes_encode_decode.c)
=============================================

```
libspdm_return_t libspdm_decode_secured_message(
--
--
--
--
    if (sequence_number > 0) {
        data64 = libspdm_read_uint64((const uint8_t *)salt) ^
                 (sequence_number - 1) ^ sequence_number;
        libspdm_write_uint64(salt, data64);
    }
--
--
}
```

As a result the salt/IV produced results to incorrect values which causes failures from device (being as responder)

Thanks,
Manish

### PR DESCRIPTION
**Pull request by steven-bellock**
Fix #2166.

Signed-off-by: Steven Bellock <sbellock@nvidia.com>

### CODE REVIEW

**Comment by steven-bellock** [comment]
@jyao1 since re-try involves re-establishing sessions I've left it up to the Requester's Integrator to perform the retry. So for now they would
1. Attempt to establish session.
2. Session fails due to endianness mismatch.
3. Integrator changes endianness.
4. Attempt to re-establish session.

In particular the Responder is not required to send `DecryptError` back to the Requester, so this might be tough to automate entirely in libspdm.

**Comment by jyao1** [comment]
Althougth auto detection is hard for the requester, the smart responder may choose to decrypt (little endian) then try decrypt (big endian), and cache the right endianness, right?

File: library/spdm_secured_message_lib/libspdm_secmes_encode_decode.c [hunk 1]
```diff
@@ -397,8 +404,15 @@ libspdm_return_t libspdm_decode_secured_message(
...
+    /* Form the AEAD IV from the salt and the sequence number. */
     libspdm_copy_mem (iv, LIBSPDM_MAX_AEAD_IV_SIZE, salt, aead_iv_size);
-    data64 = libspdm_read_uint64((const uint8_t *)iv) ^ sequence_number;
+
+    if (secured_message_context->sequence_number_endian ==
+        LIBSPDM_DATA_SESSION_SEQUENCE_NUMBER_ENDIAN_BIG) {
+        data64 = libspdm_read_uint64((const uint8_t *)iv) ^ libspdm_le_to_be_64(sequence_number);
+    } else {
+        data64 = libspdm_read_uint64((const uint8_t *)iv) ^ sequence_number;
+    }
```
Comment (jyao1) [inline library/spdm_secured_message_lib/libspdm_secmes_encode_decode.c:415]: I think we need support dual_decode, like how we support dual_verify.

For encoder, choose one of little and big, like how we support asym_sign (no change is required here)

**Comment by steven-bellock** [comment]
> Althougth auto detection is hard for the requester, the smart responder may choose to decrypt (little endian) then try decrypt (big endian), and cache the right endianness, right?

Yeah it is a cheap operation. I'll add that.

File: library/spdm_secured_message_lib/libspdm_secmes_encode_decode.c [hunk 2]
```diff
@@ -127,8 +127,15 @@ libspdm_return_t libspdm_encode_secured_message(
         return LIBSPDM_STATUS_SEQUENCE_NUMBER_OVERFLOW;
     }
 
+    /* Form the AEAD IV from the salt and the sequence number. */
     libspdm_copy_mem (iv, LIBSPDM_MAX_AEAD_IV_SIZE, salt, aead_iv_size);
-    data64 = libspdm_read_uint64((const uint8_t *)iv) ^ sequence_number;
+
+    if (secured_message_context->sequence_number_endian ==
+        LIBSPDM_DATA_SESSION_SEQUENCE_NUMBER_ENDIAN_BIG) {
+        data64 = libspdm_read_uint64((const uint8_t *)iv) ^ libspdm_le_to_be_64(sequence_number);
```
Comment (jyao1) [inline library/spdm_secured_message_lib/libspdm_secmes_encode_decode.c:135]: If the intention is to simulate what TLS does.

Then I think we should XOR IV's most significant bit, not least significant bit.

See the picture at https://github.com/DMTF/libspdm/issues/2166#issuecomment-1608838324

**Comment by jyao1** [comment]
@steven-bellock , reminder to resolve conflict.

File: library/spdm_secured_message_lib/libspdm_secmes_encode_decode.c (hunk 2 above)
Comment (steven-bellock) [reply library/spdm_secured_message_lib/libspdm_secmes_encode_decode.c:135]: Yes, that needs to be fixed.

**Comment by steven-bellock** [comment]
Still need unit tests for big-endian and try-fail. May do that in a separate pull request.

File: include/library/spdm_common_lib.h [hunk 3]
```diff
@@ -218,6 +220,12 @@ typedef enum {
 #define LIBSPDM_SPDM_10_11_VERIFY_SIGNATURE_ENDIAN_LITTLE_ONLY     (1)
 #define LIBSPDM_SPDM_10_11_VERIFY_SIGNATURE_ENDIAN_BIG_OR_LITTLE   (2)
 
+typedef enum {
+    LIBSPDM_DATA_SESSION_SEQ_NUM_ENC_LITTLE_DEC_LITTLE,
+    LIBSPDM_DATA_SESSION_SEQ_NUM_ENC_BIG_DEC_BIG,
+    LIBSPDM_DATA_SESSION_SEQ_NUM_ENC_LITTLE_DEC_BOTH,
+    LIBSPDM_DATA_SESSION_SEQ_NUM_ENC_BIG_DEC_BOTH
+} libspdm_sequence_number_endian_t;
```
Comment (jyao1) [inline include/library/spdm_common_lib.h:228]: Can we use consistent way to handle session endianness and signature endianness?

I am ok with either way - enum or uint8_t.

Comment (steven-bellock) [reply include/library/spdm_common_lib.h:228]: Sure, I can change it to `uint8_t` and use macros.

File: library/spdm_secured_message_lib/libspdm_secmes_encode_decode.c [hunk 4]
```diff
@@ -6,6 +6,62 @@
 
 #include "internal/libspdm_secured_message_lib.h"
 
+static void generate_iv(uint64_t sequence_number, uint8_t *iv, const uint8_t *salt,
```
Comment (steven-bellock) [inline library/spdm_secured_message_lib/libspdm_secmes_encode_decode.c:9]: @jyao1 please check that this makes sense.

Comment (jyao1) [reply library/spdm_secured_message_lib/libspdm_secmes_encode_decode.c:9]: looks good to me.

@manishc88, would you please double check?

File: include/library/spdm_common_lib.h [hunk 5]
```diff
@@ -214,10 +216,14 @@ typedef enum {
...
 /* The Endian values apply only if the spdm version is 1.0/1.1.
  * The default verification mode is big endian only. */
-#define LIBSPDM_SPDM_10_11_VERIFY_SIGNATURE_ENDIAN_BIG_ONLY        (0)
-#define LIBSPDM_SPDM_10_11_VERIFY_SIGNATURE_ENDIAN_LITTLE_ONLY     (1)
-#define LIBSPDM_SPDM_10_11_VERIFY_SIGNATURE_ENDIAN_BIG_OR_LITTLE   (2)
-
+#define LIBSPDM_SPDM_10_11_VERIFY_SIGNATURE_ENDIAN_BIG_ONLY 0
+#define LIBSPDM_SPDM_10_11_VERIFY_SIGNATURE_ENDIAN_LITTLE_ONLY 1
+#define LIBSPDM_SPDM_10_11_VERIFY_SIGNATURE_ENDIAN_BIG_OR_LITTLE 2
+
```
Comment (jyao1) [inline include/library/spdm_common_lib.h:222]: /* The Endian values apply only if the secured spdm version is 1.0/1.1.
 * The default sequence number mode is enc little and dec little endian only. */

File: library/spdm_secured_message_lib/libspdm_secmes_encode_decode.c (hunk 4 above)
Comment (steven-bellock) [reply library/spdm_secured_message_lib/libspdm_secmes_encode_decode.c:9]: Right now the default behavior is, for both 277 1.0 and 1.1, that little endian is used exclusively. If the Integrator wants something different then they'll need to set it in the SPDM context.

File: include/library/spdm_common_lib.h [hunk 6]
```diff
@@ -212,12 +214,20 @@ typedef enum {
...
+/* Set the endianness of the AEAD sequence number.
```
Comment (jyao1) [inline include/library/spdm_common_lib.h:223]: please clarify this is for 1.0 and 1.1 only.

**Review by jyao1 — APPROVED** [review]

**Comment by jyao1** [comment]
@steven-bellock , when the test case will be ready?

**Comment by steven-bellock** [comment]
> @steven-bellock , when the test case will be ready?

Hopefully by the end of the week. Feel free to merge this and I'll obviously fix bugs if the tests find any.

**Pull request merged**